File: python/xfa/06_pagination_multipage/06_pagination_multipage.py
# ============================================================================
#  06_pagination_multipage -- Python (ctypes) port of
#  examples\delphi\xfa\06_pagination_multipage\06_pagination_multipage.dpr
#
#  LumasPDF XFA dynamic engine "flavor tour" example 6 of 10 --
#  MULTI-PAGE PAGINATION: pageSet/pageArea/contentArea, forced overflow of a
#  70-row "Invoice Line Items" table across several pages, leader/trailer
#  "continued" banner subforms via <overflow leader=... trailer=...>.
#
#    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
#    pdfCreateXFAStreamA('datasets',...) -> pdfXFAFormPageCount (pre-flight)
#    -> pdfRenderXFAForm -> pdfCloseFile
#
#  CheckPageCount=4: hand-derived in the Delphi flavor's README.md from this
#  fixture's own geometry (contentArea 400pt tall, row/leader/trailer h=20pt
#  each, 70 <Line> records) using the engine's real pagination algorithm
#  (trailer height reserved on every page unconditionally; leader height
#  reserved on every page except the first). This driver calls
#  pdfXFAFormPageCount BEFORE pdfRenderXFAForm and asserts the two agree.
#
#  Packets are read pre-split from their own .template.xml/.datasets.xml
#  files -- no XML library needed in this driver.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

HERE = os.path.dirname(os.path.abspath(__file__))

CHECK_PAGE_COUNT = 4


def render(template_path, datasets_path, out_pdf_path, check_page_count):
    template = open(template_path, "rb").read()
    datasets = open(datasets_path, "rb").read()
    print("template packet bytes:", len(template))
    print("datasets packet bytes:", len(datasets))

    pdf = L.pdfNewPDF()
    if not pdf:
        print("pdfNewPDF FAILED")
        return -100
    try:
        if L.pdfCreateNewPDFA(pdf, out_pdf_path.encode("latin-1")) == 0:
            print("pdfCreateNewPDFA FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"template", template, len(template))
        print("pdfCreateXFAStreamA(template) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(template) FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"datasets", datasets, len(datasets))
        print("pdfCreateXFAStreamA(datasets) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(datasets) FAILED")
            return -100

        pre = L.pdfXFAFormPageCount(pdf)
        print("pdfXFAFormPageCount (pre-flight, before any AppendPage) ->", pre)
        if pre != check_page_count:
            print(f"PAGECOUNT-MISMATCH: expected {check_page_count} got {pre}")
            return -102

        rc = L.pdfRenderXFAForm(pdf)
        print("pdfRenderXFAForm ->", rc)
        if rc < 0:
            print("pdfRenderXFAForm FAILED, code", rc)
            return rc
        if rc != check_page_count:
            print(f"RENDER-PAGECOUNT-MISMATCH: pre-flight said {check_page_count} but render produced {rc}")
            return -103

        if L.pdfCloseFile(pdf) == 0:
            print("pdfCloseFile FAILED")
            return -101

        print("OK: wrote", out_pdf_path)
        return rc
    finally:
        L.pdfDeletePDF(pdf)


def main():
    template_path = os.path.join(HERE, "06_pagination_multipage.template.xml")
    datasets_path = os.path.join(HERE, "06_pagination_multipage.datasets.xml")
    out_pdf_path = os.path.join(HERE, "06_pagination_multipage.pdf")

    print("=== 06_pagination_multipage -> 06_pagination_multipage.pdf ===")
    rc = render(template_path, datasets_path, out_pdf_path, CHECK_PAGE_COUNT)
    print("RESULT|06_pagination_multipage=", rc)

    print()
    print("Expected page-by-page layout (hand-derived, verified against the")
    print("Delphi flavor's content-stream inspection via pypdf):")
    print("  page 1: header yes, leader no,  trailer yes, 19 Description rows (1-19)")
    print("  page 2: header no,  leader yes, trailer yes, 18 Description rows (20-37)")
    print("  page 3: header no,  leader yes, trailer yes, 18 Description rows (38-55)")
    print("  page 4: header no,  leader yes, trailer no,  15 Description rows (56-70)")
    print("  19 + 18 + 18 + 15 = 70 total line items, matching the 70 authored records.")


if __name__ == "__main__":
    main()

File: c/xfa/06_pagination_multipage/06_pagination_multipage.c
/* 06_pagination_multipage -- C port of examples\delphi\xfa\06_pagination_multipage
   XFA "flavor tour" example 6 of 10 -- MULTI-PAGE PAGINATION:
   pageSet/pageArea/contentArea, forced overflow of a 70-row repeating
   table across several pages, leader/trailer "continued" banner subforms
   via <overflow leader=... trailer=...>.

   Reads the pre-split packet files 06_pagination_multipage.template.xml /
   06_pagination_multipage.datasets.xml and renders them through the real
   LumasPdf.dll:

     pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
     pdfCreateXFAStreamA('datasets',...) -> pdfXFAFormPageCount (pre-flight)
     -> pdfRenderXFAForm -> pdfCloseFile

   Mirrors 06_pagination_multipage.dpr's call sequence exactly, including
   the pre-flight pdfXFAFormPageCount check asserted against the
   hand-derived expected page count of 4 (see README.md for the full
   greedy-simulation derivation: first page capacity 19 rows, each
   continuation page 18 rows, 19+18+18+15=70). Does not rebuild
   LumasPdf.dll -- links only against the public wrapper header lumaspdf.h
   and loads whatever LumasPdf.dll sits next to this exe. */
#include <stdio.h>
#include <stdlib.h>
#include "../../_common.h"
#include "lumaspdf.h"

#define EXPECTED_PAGE_COUNT 4

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    if (ErrMessage) printf("%s\n", ErrMessage);
    return -1;
}

static void* ReadWholeFile(const char* path, long* len)
{
    FILE* f; void* buf; long sz;
    f = fopen(path, "rb");
    if (!f) return NULL;
    fseek(f, 0, SEEK_END);
    sz = ftell(f);
    fseek(f, 0, SEEK_SET);
    buf = malloc(sz > 0 ? (size_t)sz : 1);
    if (!buf) { fclose(f); return NULL; }
    if (sz > 0 && fread(buf, 1, (size_t)sz, f) != (size_t)sz) { fclose(f); free(buf); return NULL; }
    fclose(f);
    *len = sz;
    return buf;
}

int main(int argc, char** argv)
{
    char dir[1024], templatePath[1100], datasetsPath[1100], outPath[1100];
    void *templateBuf, *datasetsBuf;
    long templateLen = 0, datasetsLen = 0;
    PPDF pdf;
    SI32 idx, rc, pre;
    int allPass = 1;

    exedir(argv[0], dir, sizeof(dir));
    _snprintf(templatePath, sizeof(templatePath), "%s\\06_pagination_multipage.template.xml", dir);
    _snprintf(datasetsPath, sizeof(datasetsPath), "%s\\06_pagination_multipage.datasets.xml", dir);
    _snprintf(outPath, sizeof(outPath), "%s\\06_pagination_multipage.pdf", dir);

    printf("=== 06_pagination_multipage (pre-split packets) -> 06_pagination_multipage.pdf ===\n");

    templateBuf = ReadWholeFile(templatePath, &templateLen);
    if (!templateBuf) { printf("FILE-NOT-FOUND: %s\n", templatePath); return 1; }
    datasetsBuf = ReadWholeFile(datasetsPath, &datasetsLen);
    if (!datasetsBuf) { printf("FILE-NOT-FOUND: %s\n", datasetsPath); free(templateBuf); return 1; }

    printf("template packet bytes: %ld\n", templateLen);
    printf("datasets packet bytes: %ld\n", datasetsLen);

    pdf = pdfNewPDF();
    if (!pdf) { printf("pdfNewPDF FAILED\n"); return 1; }
    pdfSetOnErrorProc(pdf, 0, ErrProc);

    if (!pdfCreateNewPDFA(pdf, outPath)) {
        printf("pdfCreateNewPDFA FAILED\n");
        pdfDeletePDF(pdf); return 1;
    }

    idx = pdfCreateXFAStreamA(pdf, "template", templateBuf, (UI32)templateLen);
    printf("pdfCreateXFAStreamA(template) -> index %d\n", idx);
    if (idx < 0) { printf("pdfCreateXFAStreamA(template) FAILED\n"); pdfDeletePDF(pdf); return 1; }

    idx = pdfCreateXFAStreamA(pdf, "datasets", datasetsBuf, (UI32)datasetsLen);
    printf("pdfCreateXFAStreamA(datasets) -> index %d\n", idx);
    if (idx < 0) { printf("pdfCreateXFAStreamA(datasets) FAILED\n"); pdfDeletePDF(pdf); return 1; }

    pre = pdfXFAFormPageCount(pdf);
    printf("pdfXFAFormPageCount (pre-flight, before any AppendPage) -> %d\n", pre);
    if (pre != EXPECTED_PAGE_COUNT) {
        printf("PAGECOUNT-MISMATCH: expected %d got %d\n", EXPECTED_PAGE_COUNT, pre);
        allPass = 0;
    }

    rc = pdfRenderXFAForm(pdf);
    printf("pdfRenderXFAForm -> %d\n", rc);
    if (rc < 1) { printf("pdfRenderXFAForm FAILED, code %d\n", rc); pdfDeletePDF(pdf); return 1; }
    if (rc != EXPECTED_PAGE_COUNT) {
        printf("RENDER-PAGECOUNT-MISMATCH: pre-flight said %d but render produced %d\n", EXPECTED_PAGE_COUNT, rc);
        allPass = 0;
    }

    if (!pdfCloseFile(pdf)) { printf("pdfCloseFile FAILED\n"); pdfDeletePDF(pdf); return 1; }
    printf("OK: wrote %s (%d page(s))\n", outPath, rc);

    pdfDeletePDF(pdf);
    free(templateBuf); free(datasetsBuf);

    printf("\nRESULT|06_pagination_multipage=%d|%s\n", rc, allPass ? "PASS" : "FAIL");
    printf("Expected checklist (see README.md):\n");
    printf("  Page 0: header yes, leader no,  trailer yes, 19 rows (1-19)\n");
    printf("  Page 1: header no,  leader yes, trailer yes, 18 rows (20-37)\n");
    printf("  Page 2: header no,  leader yes, trailer yes, 18 rows (38-55)\n");
    printf("  Page 3: header no,  leader yes, trailer no,  15 rows (56-70)\n");
    printf("  Total pages: 4  (19+18+18+15=70, matches the 70 authored <Line> records)\n");
    return allPass ? 0 : 1;
}
